src/main/java/org/eclipse/uprotocol/communication/InMemorySubscriber.java

```diff
@@ -66,31 +92,51 @@ public InMemorySubscriber (UTransport transport, RpcClient rpcClient) {
+     * to the subscription state (ex. PENDING to SUBSCRIBED, or SUBSCRIBED to UNSUBSCRIBED, etc...).
      * 
      * @param topic The topic to subscribe to.
      * @param listener The listener to be called when a message is received on the topic.
      * @param options The call options for the subscription.
-     * @return Returns the CompletionStage with the response UMessage or exception with the failure
+     * @param handler {@link SubscriptionChangeHandler} to handle changes to subscription states.
+     * @return Returns the CompletionStage with {@link SubscriptionResponse} or exception with the failure
      * reason as {@link UStatus}.
      */
     @Override
-    public CompletionStage<SubscriptionResponse> subscribe(UUri topic, UListener listener, CallOptions options) {
+    public CompletionStage<SubscriptionResponse> subscribe(UUri topic, UListener listener, CallOptions options,
+        SubscriptionChangeHandler handler) {
         Objects.requireNonNull(topic, "Subscribe topic missing");
         Objects.requireNonNull(listener, "Request listener missing");
+        options = Objects.requireNonNullElse(options, CallOptions.DEFAULT);
 
-        final UUri subscribe = UriFactory.fromProto(
-            USubscriptionProto.getDescriptor().getServices().get(0), METHOD_SUBSCRIBE); 
-        
         final SubscriptionRequest request = SubscriptionRequest.newBuilder()
             .setTopic(topic)
             .setSubscriber(SubscriberInfo.newBuilder().setUri(transport.getSource()).build())
             .build();
         
+        // Send the subscription request and handle the response
         return RpcMapper.mapResponse(rpcClient.invokeMethod(
-                subscribe, UPayload.pack(request), options), SubscriptionResponse.class)
-            .thenApplyAsync(response -> {
-                transport.registerListener(topic, listener).toCompletableFuture().join();
+                SUBSCRIBE_METHOD, UPayload.pack(request), options), SubscriptionResponse.class)
+            
+            // Then register the listener to be called when messages are received
+            .thenCompose(response -> {
+                if ( response.getStatus().getState() == SubscriptionStatus.State.SUBSCRIBED ||
+                    response.getStatus().getState() == SubscriptionStatus.State.SUBSCRIBE_PENDING) {
+                    return transport.registerListener(topic, listener).thenApply(status -> response);
+                }
+                return CompletableFuture.completedFuture(response);
+            })
+
+            // Then Add the handler (if present) to the hashMap
+            .thenApply(response -> {
+                if (handler != null) {
+                    mHandlers.compute(topic, (key, currentHandler) -> {
+                        if (currentHandler != null) {
+                            throw new UStatusException(UCode.ALREADY_EXISTS, "Handler already registered");
```

Comment:
This seems to mean that the uE has already been subscribed to the topic, right? If I am not mistaken, then throwing this exception here will result in the uE assuming that subscribing to the topic did not work, when in fact it (still) is subscribed to the topic, right?

Reply:
Ok I fixed it so that the call to subscribe is idempotent (if you pass the same information multiple times) since that is how uSubscription service works.